R/utils.R

```diff
@@ -471,37 +473,66 @@ remapListVar <- function(map, listVarPlotRef, newValuePlotRef, newVarId = 'melte
+    stop("listVar error: all vars in a listVar must have the same entity id, type, and shape.")
   }
-  
-  # Ensure repeatedPlotRef is numeric
-  if (any(map$dataType[map$plotRef == listVarPlotRef] != 'NUMBER')) {
-    stop(paste0("All vars in ", listVarPlotRef, " must be of type NUMBER."))
+
+  # Ensure all variables are numbers
+  if (!all(listVariable$dataType == 'NUMBER')){
+    stop("listVar error: All vars must be of type NUMBER.")
   }
 
-  # Check to ensure if repeatedPlotRef is facet that there are no other facet vars.
-  if (listVarPlotRef == 'facetVariable1' & 'facetVariable2' %in% map$plotRef) {
-    stop("facetVariable2 should be NULL when using listVar for facetVariable1.")
+  # Ensure all variables are continuous
+  if (!all(listVariable$dataShape == 'CONTINUOUS')){
+    stop("listVar error: All vars must be CONTINUOUS.")
   }
 
-  # Check we do not have too many vars -- restricted based on the rules in the data service
-  nVars <- length(map$id[map$plotRef == listVarPlotRef])
-  if (listVarPlotRef == 'xAxisVariable' & nVars > 10) {
-    stop("Too many values specified with listVar: maximum number of x axis values is 10.")
-  } else if (listVarPlotRef == 'overlayVariable' & nVars > 8) {
-    stop("Too many values specified with listVar: maximum number of overlay values is 8.")
-  } else if (listVarPlotRef == 'facetVariable1' & nVars > 25) {
-    stop("Too many values specified with listVar: maximum number of panels allowed is 25.")
+  # Ensure no to variables are the same
+  if (any(duplicated(listVariable$variableId))) {
+    stop("listVar error: No duplicate vars allowed.")
   }
-  
-  # Require all repeated vars to have the same type, shape, and entity
-  if (uniqueN(map$entityId[map$plotRef == listVarPlotRef]) > 1 | uniqueN(map$dataType[map$plotRef == listVarPlotRef]) > 1 | uniqueN(map$dataShape[map$plotRef == listVarPlotRef]) > 1) {
-    stop("All vars in a listVar must have the same entity id, type, and shape.")
+
+  return(listVariable)
+}
+
+validateMap <- function(map) {
+  # Could add checks for data type, shape, etc presence
+
+  if (any(duplicated(map$plotRef))) {
+    
+    # Check that there is at most one repeated plotRef
+    repeatedPlotRef <- unique(map$plotRef[duplicated(map$plotRef)])
+
+    if (length(repeatedPlotRef) > 1) {
+      stop("map error: only one plotRef can be assigned to multiple variables.")
+    }
+
+    # Check we do not have too many repeated variables.
+    nVars <- sum(duplicated(map$plotRef))
+
+    if (repeatedPlotRef == 'xAxisVariable' & nVars > 10) {
+      stop("Too many values specified with listVar: maximum number of x axis values is 10.")
+    } else if (repeatedPlotRef == 'overlayVariable' & nVars > 8) {
+      stop("Too many values specified with listVar: maximum number of overlay values is 8.")
+    } else if (repeatedPlotRef == 'facetVariable1' & nVars > 10) {
+      stop("Too many values specified with listVar: maximum number of panels allowed is 10.")
```

Comment:
should double check the rules used in the data service for facets.. i think we said the total number of panels could be larger than 10, and we dont limit based on individual facet vars. also this conditions tests on a specific facet, and then the message talks about panels which are the interactions between both facets.